### PR TITLE
[ATen batch norm]

### PR DESCRIPTION
**Pull request by jjsjann123**
fix cuda native batch norm for small feature planes.
  1. fixed warp reduction divergent call of WARP_SHFL_XOR, causes hang with CUDA_ARCH > 7.0
  2. split Normalization.cu into two files for code reuse, preparation for sync BN

### CODE REVIEW

File: aten/src/ATen/native/cuda/Normalization.cu [hunk 1]
```diff
@@ -1,516 +1,7 @@
...
-      var_n += d1 * (v - avg);
-    }
-  }
-
-  // first warpSum to get one value per thread to
-  // one value per warp
-  for (int i = 0; i < getMSB(WARP_SIZE); ++i) {
-    accscalar_t o_avg = WARP_SHFL_XOR(avg, 1 << i, WARP_SIZE);
-    int o_n = WARP_SHFL_XOR(n, 1 << i, WARP_SIZE);
-    if (n + o_n > 0) {
```
Comment (jjsjann123) [inline aten/src/ATen/native/cuda/Normalization.cu]: This is the real code change, not all threads in warp calls WARP_SHFL_XOR causes cuda hange on volta.

File: aten/src/ATen/native/cuda/Normalization.cu [hunk 2]
```diff
@@ -1,516 +1,7 @@
...
-
-  if (tid < WARP_SIZE) {
-    n = (tid < blockDim.x * blockDim.y / WARP_SIZE ? shared_n[tid] : 0);
-    avg = (tid < blockDim.x * blockDim.y  / WARP_SIZE ? shared_avg_var[2 * tid] : 0);
-    var_n = (tid < blockDim.x * blockDim.y  / WARP_SIZE ? shared_avg_var[2 * tid + 1] : 0);
-  }
-  for (int i = 0; i < getMSB(WARP_SIZE); ++i) {
-    accscalar_t o_avg = WARP_SHFL_XOR(avg, 1 << i, WARP_SIZE);
-    int o_n = WARP_SHFL_XOR(n, 1 << i, WARP_SIZE);
-    if (n + o_n > 0) {
```
Comment (jjsjann123) [inline aten/src/ATen/native/cuda/Normalization.cu]: Same as above

File: aten/src/ATen/native/cuda/Normalization.cuh [hunk 3]
```diff
@@ -0,0 +1,511 @@
...
+      var_n += d1 * (v - avg);
+    }
+  }
+
+  // first warpSum to get one value per thread to
+  // one value per warp
+  for (int i = 0; i < getMSB(WARP_SIZE); ++i) {
+    accscalar_t o_avg = WARP_SHFL_XOR(avg, 1 << i, WARP_SIZE);
+    int o_n = WARP_SHFL_XOR(n, 1 << i, WARP_SIZE);
+    accscalar_t factor = 1.0 / fmaxf(1.0, n+o_n);
```
Comment (jjsjann123) [inline aten/src/ATen/native/cuda/Normalization.cuh]: Removed the divergent calls and let non-contributing threads participate in warp sync calls

**Review by t-vi — APPROVED** [review]
I'm not sure about splitting the file (is that for sync batch norm?), but if the changes beyond the split are what @jjsjann123 points to, I'd love to see my mistakes fixed.
Did you check the CI?

**Review by soumith — APPROVED** [review]

**Comment by jjsjann123** [comment]
I tried the fixes with test_nn.py. It addressed the hang I was seeing before the fixes.
Looking at the CI, it seems to be failing at backward reentrant in test_spectral_norm. Not sure if it's related to my changes.

**Comment by soumith** [comment]
CI issue is not related to your PR, I am landing it.

**Pull request closed without merging**